src/Control/Monad/Free.purs

```diff
@@ -1,245 +1,233 @@
 import Data.Eq (class Eq1, eq1)
+import Data.Exists (Exists, mkExists, runExists)
 import Data.Foldable (class Foldable, foldMap, foldl, foldr)
-import Data.Maybe (Maybe(..))
 import Data.Ord (class Ord1, compare1)
 import Data.Traversable (class Traversable, traverse)
-import Data.Tuple (Tuple(..))
-
 import Unsafe.Coerce (unsafeCoerce)
 
--- | The free monad for a type constructor `f`.
--- |
--- | Implemented in the spirit of [Reflection without Remorse](http://okmij.org/ftp/Haskell/zseq.pdf),
--- | the free monad is represented using a sequential data structure in
--- | order to overcome the quadratic complexity of left-associated binds
--- | and traversal through the free monad structure.
-data Free f a = Free (FreeView f Val Val) (CatList (ExpF f))
-
-newtype ExpF f = ExpF (Val -> Free f Val)
+foreign import data UnsafeBoundValue :: Type
 
-data FreeView f a b = Return a | Bind (f b) (b -> Free f a)
+foreign import data UnsafeBoundF :: Type -> Type
 
-data Val
+data Free f a
+  = Pure a
+  | Bind (f UnsafeBoundValue) (FreeBinds f UnsafeBoundValue a)
 
 instance eqFree :: (Functor f, Eq1 f, Eq a) => Eq (Free f a) where
-  eq x y = case resume x, resume y of
+  eq x y = case resume' x, resume' y of
     Left fa, Left fb -> eq1 fa fb
-    Right a, Right b -> a == b
+    Right a, Right b -> eq a b
     _, _ -> false
 
-instance eq1Free :: (Functor f, Eq1 f) => Eq1 (Free f) where
-  eq1 = eq
-
 instance ordFree :: (Functor f, Ord1 f, Ord a) => Ord (Free f a) where
-  compare x y = case resume x, resume y of
+  compare x y = case resume' x, resume' y of
     Left fa, Left fb -> compare1 fa fb
     Left _, _ -> LT
     _, Left _ -> GT
     Right a, Right b -> compare a b
 
+instance eq1Free :: (Functor f, Eq1 f) => Eq1 (Free f) where
+  eq1 = eq
+
 instance ord1Free :: (Functor f, Ord1 f, Ord a) => Ord1 (Free f) where
   compare1 = compare
 
-instance freeFunctor :: Functor (Free f) where
-  map k f = pure <<< k =<< f
+instance functorFree :: Functor (Free f) where
+  map f (Pure a) = Pure (f a)
+  map f (Bind a bs) = Bind a (Node (unsafeCoerce bs) (Leaf (Pure <<< unsafeCoerce f)))
 
-instance freeBind :: Bind (Free f) where
-  bind (Free v s) k = Free v (snoc s (ExpF (unsafeCoerceBind k)))
-    where
-    unsafeCoerceBind :: forall a b. (a -> Free f b) -> Val -> Free f Val
-    unsafeCoerceBind = unsafeCoerce
+instance applyFree :: Apply (Free f) where
+  apply = ap
 
-instance freeApplicative :: Applicative (Free f) where
-  pure = fromView <<< Return
+instance applicativeFree :: Applicative (Free f) where
+  pure = Pure
 
-instance freeApply :: Apply (Free f) where
-  apply = ap
+instance bindFree :: Bind (Free f) where
+  bind (Pure a) k = k a
+  bind (Bind a bs) k = Bind a (Node (unsafeCoerce bs) (Leaf (unsafeCoerce k)))
 
-instance freeMonad :: Monad (Free f)
+instance monadFree :: Monad (Free f)
 
-instance freeMonadTrans :: MonadTrans Free where
-  lift = liftF
+instance monadTransFree :: MonadTrans Free where
+  lift = lift
```

Comment:
@xgrommx The `lift` in this library is the same as the `lift` from `MonadTrans` -- it's one reason why I think it's sensible to rename the function.